### PR TITLE
OIEWP-1041 Feature caching

### PR DESCRIPTION
**Pull request by ollie-sims**
Added feature caching for the IMS layer. Caching can be switched on or off using the property enableCache. Features can be cached for the last n number of map bounds. The max limit of cached bounds is defined by the property maxBoundsCacheSize.
The cache consists of an array of map bounds, a set of feature ids stored and an rtree of features and there bounds. The bounds cache is checked after a map change to see if the features can be retrieved from the cache. If the new bounds is covered by the cache, a feature collection is created from the cache and used to populate the layer using the existing code and no requests are made to the IMS service.
Replaced clearHighlights() with redrawIMSLayer().
Updated _clearIMSLayer() to remove any image load functions to prevent any out of sync rendering of points when the map has changed.
In the interest of getting some progress here, units tests for the caching will follow in a future PR.

### CODE REVIEW

File: px-map-behavior-layer-IMS.es6.js [hunk 1]
```diff
@@ -191,13 +191,376 @@
...
+      return !(otherBounds.minX >= bounds.maxX ||
+        otherBounds.minY >= bounds.maxY ||
+        otherBounds.maxX <= bounds.minX ||
+        otherBounds.maxY <= bounds.minY);
+    },
+
+    /**
+     * Returns whether otherBounds contains bounds.
+     */
+    _boundsContains(bounds, otherBounds) {
```
Comment (Dave-Fox) [inline px-map-behavior-layer-IMS.es6.js:270]: This is the same as `_boundsInteracts` ? Can we remove one of them?

File: px-map-behavior-layer-IMS.es6.js [hunk 2]
```diff
@@ -191,13 +191,376 @@
...
+      return !(otherBounds.minX > bounds.minX ||
+        otherBounds.minY > bounds.minY ||
+        otherBounds.maxX < bounds.maxX ||
+        otherBounds.maxY < bounds.maxY);
+    },
+
+    /**
+     * Returns an array of bounds defining the area of bounds that is not overlapped by otherBounds.
+     */
+    _splitBounds(bounds, otherBounds) {
```
Comment (Dave-Fox) [inline px-map-behavior-layer-IMS.es6.js]: This function really needs to be broken down to make sense, perhaps move the scope of the top-level `if`, `else if` and `else` to live in their own functions as a first step.

File: px-map-behavior-layer-IMS.es6.js [hunk 3]
```diff
@@ -191,13 +191,376 @@
...
+    },
+
+    /**
+     * Returns whether testBounds is overlapped by all of the bounds in boundsArray.
+     */
+    _boundsCovered(testBounds, boundsArray) {
+      if (!boundsArray) {
+        boundsArray = this._boundsCache;
+      }
+      for (let b of boundsArray) {
```
Comment (Dave-Fox) [inline px-map-behavior-layer-IMS.es6.js]: `b` -> `bound` so that it is meaningful

File: px-map-behavior-layer-IMS.es6.js [hunk 4]
```diff
@@ -191,13 +191,376 @@
...
+        if (this._boundsContains(testBounds, b)) {
+          return true;
+        }
+      }
+      const n = boundsArray.length;
```
Comment (Dave-Fox) [inline px-map-behavior-layer-IMS.es6.js]: `n` -> `boundsLength` or similar

File: px-map-behavior-layer-IMS.es6.js [hunk 5]
```diff
@@ -191,13 +191,376 @@
...
+      for (let i = 0; i < n; i++) {
+        const bounds = boundsArray[i];
+        if (this._boundsOverlaps(testBounds, bounds)) {
+          const splits = this._splitBounds(testBounds, bounds);
+          const newBoundsArray = boundsArray.slice();
+          newBoundsArray.splice(i, 1); // remove bounds for efficiency
+          let covered = true;
```
Comment (Dave-Fox) [inline px-map-behavior-layer-IMS.es6.js]: Move this `let` next to the `const`s and then add a blank line for readability

File: px-map-behavior-layer-IMS.es6.js [hunk 6]
```diff
@@ -191,13 +191,376 @@
...
+        maxY: mapBounds._northEast.lat
+      };
+    },
+
+    /**
+     * Returns an array of position pairs from the feature coordinates.
+     */
+    _getFeatureCoords(feature) {
+      let coords = [];
+      const featureCoords = feature.geometry.coordinates;
```
Comment (Dave-Fox) [inline px-map-behavior-layer-IMS.es6.js:415]: Add a blank line after this.

File: px-map-behavior-layer-IMS.es6.js [hunk 7]
```diff
@@ -191,13 +191,376 @@
...
+        return;
+      }
+
+      const mapBounds = this._getMapBounds();
+
+      /*
+       * Remove older bounds if exceeding max cache size - but only if they don't
+       * overlap with the current bounds.
+       */
+      const n = this.maxBoundsCacheSize - 2;
```
Comment (Dave-Fox) [inline px-map-behavior-layer-IMS.es6.js]: `n` -> `boundingCacheSize` or similar

File: px-map-behavior-layer-IMS.es6.js [hunk 8]
```diff
@@ -191,13 +191,376 @@
...
+      for (let i = this._boundsCache.length - 1; i > n; i--) {
+        const bounds = this._boundsCache[i];
+        if (!this._boundsOverlaps(bounds, mapBounds)) {
+          const elementsToRemove = this._featureTreeCache.search(bounds);
+          elementsToRemove.forEach(element => {
+            let remove = true;
+            for (let b of this._boundsCache) {
```
Comment (Dave-Fox) [inline px-map-behavior-layer-IMS.es6.js]: Rename `b` to a meaningful variable name

File: px-map-behavior-layer-IMS.es6.js [hunk 9]
```diff
@@ -191,13 +191,376 @@
...
+              this._featureCache.delete(element.feature.id);
+              this._featureTreeCache.remove(element);
+            }
+          });
+          this._boundsCache.splice(i, 1);
+        }
+      }
+
+      // Add new features from the current feature collection to the cache
+      const features = this.featureCollection.features;
```
Comment (Dave-Fox) [inline px-map-behavior-layer-IMS.es6.js]: Can this block be moved to its own function? It would help reduce the size of this one.

**Review by Dave-Fox — CHANGES_REQUESTED** [review]

File: px-map-behavior-layer-IMS.es6.js (hunk 1 above)
Comment (ollie-sims) [reply px-map-behavior-layer-IMS.es6.js:270]: They are different!

**Review by Dave-Fox — APPROVED** [review]
Looks good.

**Pull request merged**